### PR TITLE
Feat: Expose export params

### LINKED ISSUE #379: Feat: Expose *ExportParams in Filters
## Background

Since my use case needs to send images to embedded devices over bluetooth, the output size is of top priority. Following advice in #20, I have found using png with `Palette` and `Bitdepth` to work great.

Currently, I'm running an instance with `NewPngExportParams()` modified. It would be great if these params are exposed in filters instead for ease of use.

## Proposal

Filter `format(png(palette=true, bitdepth=4, compression=8))`

## Implementation

Currently, the filters seems to all use positional parameters instead of named parameters. But with image export options, it would be reasonable to leave most settings as default, and only change ones that are needed. Hence I think named parameters would be more suitable. However I have no knowledge of how parsing is currently done to know if there is difficulty in implementation.

### PR DESCRIPTION
**Pull request by chakflying**
## Description

Fixes #379

Add parameters for exported file formats as optional filters.

```
format(png):palette():bitdepth(4):compression(8)
```

The code should work with PNGs as is, but I'm not entirely familiar with golang so I'm opening this as draft to see if the current idea of implementation is okay.

### CODE REVIEW

**Comment by chakflying** [comment]
Code has been running fine in production for a while. Can add tests if there is interest.

File: vips/process.go [hunk 1]
```diff
@@ -300,8 +320,16 @@ func (v *Processor) Process(
 		return imagor.NewBlobFromJsonMarshal(metadata(img, format, stripExif)), nil
 	}
 	format = supportedSaveFormat(format) // convert to supported export format
+	exportParams := &ExportParams{
```
Comment (cshum) [inline vips/process.go:323]: Seems creating another ExportParams struct is unnecessary?
Would prefer just populate extra arguments into the `export` function.

The arguments do look a bit more cumbersome I agree, but at least the changes are simpler.

Comment (chakflying) [reply vips/process.go:323]: My thinking was that there can be many more parameters that someone would like to be able to modify in future, e.g. those in `Jpeg` and `Webp`. Passing them all individually would look very messy.

Comment (cshum) [reply vips/process.go:323]: At that point it is probably better to break that down into separated functions, as they have already got their own export params.

File: vips/process.go [hunk 2]
```diff
@@ -562,45 +590,49 @@ func supportedSaveFormat(format ImageType) ImageType {
...
-func (v *Processor) export(image *Image, format ImageType, quality int) ([]byte, error) {
-	switch format {
+func (v *Processor) export(image *Image, params *ExportParams) ([]byte, error) {
+	switch params.format {
 	case ImageTypePNG:
 		opts := NewPngExportParams()
+		opts.Quality = params.Quality
+		opts.Palette = params.Palette
+		opts.Bitdepth = params.Bitdepth
+		opts.Compression = params.Compression
```
Comment (cshum) [inline vips/process.go:600]: Add a `if` condition to only populate the options if being set.
In Go `int` is default at `0` and `bool` is default at `false`

https://github.com/cshum/imagor/blob/master/vips/foreign.go#L48 
As you can see the default compression is 6. Should only override `if compression > 0`
This avoid default params being overridden, resulting breaking changes.

**Review by cshum — CHANGES_REQUESTED** [review]

**Comment by chakflying** [comment]
Modified code according to suggestions.

**Comment by cshum** [comment]
Would you try also add another line of test around https://github.com/cshum/imagor/blob/master/vips/processor_test.go#L74
with the filters `format(png):palette():bitdepth(4):compression(8)`.

The test image should be auto generated by github action if running properly.

**Review by cshum — APPROVED** [review]

**Pull request merged**